### DISCUSSION TITLE
Webots and CFLib Confusion

### DISCUSSION THREAD

**Discussion by mhl787156** [bitcraze/discussions#1187] [Q&A] [answered]
Hi Kimberly, 

I've been trying to setup a webots simulation and write some controller code which can be directly ported to running with a real crazyflie later on. Ideally that means using CFLib which magically connects up to webots somehow! However I've been looking around and there seem to be PRs which do stuff like this (e.g. https://github.com/bitcraze/crazyflie-simulation/pull/43, https://www.bitcraze.io/documentation/repository/crazyflie-simulation/main/functional_areas/controllers/#firmware-python-bindings) but none of it seems to be in the main crazyflie-simulation branch anymore! 

There is the webots branch, but it looks like it hasn't been touched in a while! (https://github.com/bitcraze/crazyflie-simulation/tree/webots-stability). So is this still supported? 

If not, what exactly is the process if sim2real? If, for instance, I want to use the cflib MotionCommander, how do I set it up so that it works in webots and on the crazyflie without having to write my own motioncommander which doesn't require a synced crazyflie instance. Also don't want the hassle of setting up ROS just yet! 

I've been finding myself bouncing between the crazyflie-simulation docs and the crazyflie-lib-python docs getting confused at how to connect to two together! Apologies if this has been asked before but I couldn't find a recent discussion or PR on this! 

Many Thanks,

Mickey Li

**Comment by knmcguire** [comment, marked as answer]
Hi! So, unfortunately,  the CFLIB and webots simulation is completely disconnected. The python api is very different and you can not switch easily between the crazyflie sim and the real one. The plans were to implement that at one point but due to illness and other priority projects at bitcraze we have not come that far. If you want that type of functionality  you'll need to implement that yourself or make an issue in the the github repo if it doesn't exist yet. You don't have to use ROS for it. TCP, zeromq, zenoh could be used for that as well. But the full infrastructure needs to be setup and that takes a lot of work. This is one of the [discussion thread ](https://github.com/orgs/bitcraze/discussions/863) that are of interest for you perhaps?

For webots there is an example where you can run the firmware controllers  though through python bindings, which you can find here in current master: https://github.com/bitcraze/crazyflie-simulation/tree/main/simulator_files/webots/controllers/crazyflie_controller_py_firmware_pid

Also you can consider the gazebo route with software in the loop. These blogpost/discussions would be interesting as they also achieved a connection with the cflib as well.
- https://www.bitcraze.io/2024/01/simulation-sim_cf2-for-multi-agent-algorithm-development/
- https://github.com/orgs/bitcraze/discussions/995

**Reply by mhl787156** [reply]
Thanks for the reply! Ah I see! I was seeing bits and pieces of the discussion but I wasn't aware of how far it had gotten. I might try and give it a go, though no guarantees of any useable outcomes haha! 

Just to clarify that the firmware controllers through python bindings is not the same as the crazyflie-lib-python cflib? Sorry I get confused between all the different layers in the system! 

One alternative is ROS with webots - is there a crazyflie/ros/webots pipeline thing floating around somewhere? Maybe ROS will end up being the quickest route. 

The gazebo route would be the most ideal since the majority of projects use it! But the team seem to be settled on using a more user friendly simulator for now.

**Reply by knmcguire** [reply]
Yes you are right that the python bindings are not related to the cflib. The python bindings are wrappers around the firmware c code, and the cflib is only the library that controls the crazyflie.

Currently, none of the examples for SITL or ROS use webots. I've worked with an [experimental project ](https://github.com/knmcguire/crazyflie_ros2_experimental) myself where I got that to work but this hasn't been intergrated to staple ROS projects like [Crazyswarm2](https://imrclab.github.io/crazyswarm2/) due to incompatibility (for now) with the framework.  But perhaps you can use it as an inspiration.
